File: sdk/typescript/src/tracer.ts
import { SDKConfig, SpanConfig } from './types';
import { Span } from './span';
import { Batcher } from './batcher';

export class TraceLens {
  private config: SDKConfig;
  private batchers: Map<string, Batcher> = new Map();

  constructor(config: SDKConfig) {
    this.config = {
      bufferMs: 100,
      maxBatchSize: 100,
      ...config,
    };
  }

  startSpan(cfg: SpanConfig): Span {
    const span = new Span(cfg);
    let batcher = this.batchers.get(cfg.traceId);
    if (!batcher) {
      batcher = new Batcher(
        this.config.endpoint,
        this.config.apiKey,
        cfg.traceId,
        this.config.bufferMs,
        this.config.maxBatchSize
      );
      this.batchers.set(cfg.traceId, batcher);
    }

    const originalEnd = span.end.bind(span);
    span.end = (endCfg) => {
      const data = originalEnd(endCfg);
      batcher!.add(data);
      return data;
    };

    return span;
  }

  async flush(): Promise<void> {
    await Promise.all(Array.from(this.batchers.values()).map(b => b.stop()));
    this.batchers.clear();
  }
}

File: sdk/rust/src/tracer.rs
use crate::types::{Config, SpanConfig};
use crate::span::Span;
use crate::batcher::Batcher;
use std::collections::HashMap;
use std::sync::Arc;

pub struct TraceLens {
    config: Config,
    batchers: Arc<tokio::sync::Mutex<HashMap<String, Arc<Batcher>>>>,
}

impl TraceLens {
    pub fn new(config: Config) -> Self {
        Self {
            config,
            batchers: Arc::new(tokio::sync::Mutex::new(HashMap::new())),
        }
    }

    pub async fn start_span(&self, config: SpanConfig) -> Span {
        let mut span = Span::new(config.clone());
        let mut batchers = self.batchers.lock().await;

        let batcher = batchers.entry(config.trace_id.clone()).or_insert_with(|| {
            Batcher::new(
                self.config.endpoint.clone(),
                self.config.api_key.clone(),
                config.trace_id.clone(),
                self.config.buffer_ms,
                self.config.max_batch_size,
            )
        }).clone();

        drop(batchers);

        // Note: In production, use a callback or channel pattern.
        // Simplified here for the example.
        span
    }

    pub async fn flush(&self) {
        let batchers = self.batchers.lock().await;
        for b in batchers.values() {
            b.stop().await;
        }
    }
}
